db/schema_applier.cc

```diff
@@ -811,30 +817,56 @@ future<> schema_applier::update() {
             _before.scylla_aggregates, _after.scylla_aggregates);
 
     co_await drop_types(_proxy, _affected_user_types);
+}
+
+void schema_applier::commit_on_shard(replica::database& db) {
+    // commit keyspace operations
+    for (auto& ks_per_shard : _affected_keyspaces.created) {
+        auto ks = ks_per_shard[this_shard_id()].release();
+        db.insert_keyspace(std::move(ks));
```

Comment:
This is atomic wrt the scheduler, but it's not atomic wrt failures. If this throws, then all following updates won't be done.

In practice we can't recover from a partial schema update. Please add an on_fatal_internal_error() to commit finalization (can be a new patch) so at least we know something bad happened.

If we really want, we can have prepare() acquire unordered_map::node_type (or whatever is relevant during prepare), and use it during commit, so it's truly atomic, but let's not touch it now.

Reply:
added  on_fatal_internal_error